### ISSUE TITLE
Error: The `onCancel` handler was attached after the promise settled.

### ISSUE THREAD

**Issue by Songkeys** [microlinkhq/html-get#147]
I was just testing against the demo. Sometimes it works, sometimes it throws this error:

```
/Users/songkeys/GitHub/Crossbell-Box/faas/node_modules/p-cancelable/index.js:48
                                        throw new Error('The `onCancel` handler was attached after the promise settled.');
           ^
Error: The `onCancel` handler was attached after the promise settled.
    at onCancel (/Users/songkeys/GitHub/Crossbell-Box/faas/node_modules/p-cancelable/index.js:48:12)
    at makeRequest (/Users/songkeys/GitHub/Crossbell-Box/faas/node_modules/got/dist/source/as-promise/index.js:38:13)
    at Request.<anonymous> (/Users/songkeys/GitHub/Crossbell-Box/faas/node_modules/got/dist/source/as-promise/index.js:143:17)
    at Object.onceWrapper (node:events:628:26)
    at Request.emit (node:events:513:28)
    at Timeout.retry (/Users/songkeys/GitHub/Crossbell-Box/faas/node_modules/got/dist/source/core/index.js:1278:30)
```

Environment: macOS m1, nodejs v18.

```
demo-app@0.0.1 /Users/songkeys/GitHub/Crossbell-Box/faas
└─┬ html-get@2.9.23
  ├─┬ got@11.8.5
  │ └── p-cancelable@2.1.1 deduped
  ├── p-cancelable@2.1.1
  └─┬ top-sites@1.1.124
    └─┬ p-any@3.0.0
      ├── p-cancelable@2.1.1 deduped
      └─┬ p-some@5.0.0
        └── p-cancelable@2.1.1 deduped
```

~~I believe this is an issue related to the `got` package. See https://github.com/sindresorhus/got/issues/1489.~~

~~I would also highly recommend deprecating the use of `got` as it's somehow buggy and slow. Instead, use the [`undici`](https://undici.nodejs.org/) package, which would be much more performant. In the future (less than a month later), when nodejs v18 comes out, there is going to be built-in standard `fetch` which is also provided by `undici`.~~

---

edit: After taking another deep look, I think I found the fix...

```diff
+  if (req._isPending) {
      onCancel(() => {
        debug('fetch:cancel', { url, reflect })
        req.cancel()
      })
+   }
```

let me know what you think.

**Comment by Kikobeats** [comment]
hum, it disturbing me a bit `onCancel` is not taking care of this 🤔

**Comment by Kikobeats** [comment]
Do you think it could be related to https://github.com/sindresorhus/p-cancelable/pull/34 ?

**Comment by Songkeys** [comment]
No. The reason is that when you are defining `onCancel`, `p-cancelable` will check the pending status and throw error if invalid:

https://github.com/sindresorhus/p-cancelable/blob/cdec148b9c5995e8f4dae05efb2cd56636d7a728/index.js#L51-L54

(The above snippet is of the latest version, while the version ^2 we are using has the same logic.)

**Comment by Kikobeats** [comment]
@Songkeys I agree, but what I was thinking if we can send a PR to p-cancellable to land this check as something internal (since it's an internal variable) to patch it for v2 version

**Comment by Songkeys** [comment]
We might need a minimal reproduction to make that PR. I'm still not so sure why https://github.com/microlinkhq/html-get/blob/943b82087dd081b5fd44ef0159845e955d9b4f85/src/index.js#L30 is ahead of https://github.com/microlinkhq/html-get/blob/943b82087dd081b5fd44ef0159845e955d9b4f85/src/index.js#L36 but is got triggered after the latter. Perhaps some version of nodejs brought into some event loop changes (cuz in https://github.com/sindresorhus/got/issues/1489 many claims they encountered this issue since node 16)?

Another option - In our code, I think we can avoid the use of p-cancellable. Because we are using `got`, and it has a `req.cancel()` method already, which I think is built upon p-cancellable anyway. After that, we will probably unfortunately found the same issue again, as many people raised in https://github.com/sindresorhus/got/issues/1489. Let's take it down again, use a `fetch` plus `AbortController` to make things more modern and less buggy without p-cancellable.

But you were right. It looks like the culprit of the issues above. If we make a PR there, many would benefit. Do you have time to do so? I could do but not now. Maybe 2 more days to have a further investigation.

**Comment by Kikobeats** [comment]
I prefer to keep `got` over `fetch` since it's widely adopted by Nodeland and `fetch` miss some features there (like caching or hooks, etc.)

**Comment by Kikobeats** [comment]
This error was fixed in got@11.8.6
https://github.com/sindresorhus/got/compare/v11.8.5...v11.8.6
